clash-lib/src/Clash/Normalize/Transformations/Case.hs

```diff
@@ -0,0 +1,673 @@
+  let (xes1,e1) = deshadowLetExpr is0 xes e
+      binds = map (second (`mkTicks` ticks)) xes1
+      scrut = mkTicks e1 ticks
+   in changed (Letrec binds (Case scrut ty alts))
+
+caseLet _ e = return e
+{-# SCC caseLet #-}
+
+caseOneAlt :: Term -> NormalizeSession Term
+caseOneAlt e@(Case _ _ [(pat,altE)]) = case pat of
+  DefaultPat -> changed altE
+  LitPat _ -> changed altE
+  DataPat _ tvs xs
+    | (coerce tvs ++ coerce xs) `localVarsDoNotOccurIn` altE
+    -> changed altE
+    | otherwise
+    -> return e
+
+caseOneAlt (Case _ _ alts@((pat,alt):_:_))
+  | all ((== alt) . snd) (tail alts)
+  , (tvs,xs) <- patIds pat
+  , (coerce tvs ++ coerce xs) `localVarsDoNotOccurIn` alt
+  = changed alt
+
+caseOneAlt e = return e
+{-# SCC caseOneAlt #-}
+
+-- | Tries to eliminate existentials by using heuristics to determine what the
+-- existential should be. For example, consider Vec:
+--
+--    data Vec :: Nat -> Type -> Type where
+--      Nil       :: Vec 0 a
+--      Cons x xs :: a -> Vec n a -> Vec (n + 1) a
+--
+-- Thus, 'null' (annotated with existentials) could look like:
+--
+--    null :: forall n . Vec n Bool -> Bool
+--    null v =
+--      case v of
+--        Nil  {n ~ 0}                                     -> True
+--        Cons {n1:Nat} {n~n1+1} (x :: a) (xs :: Vec n1 a) -> False
+--
+-- When it's applied to a vector of length 5, this becomes:
+--
+--    null :: Vec 5 Bool -> Bool
+--    null v =
+--      case v of
+--        Nil  {5 ~ 0}                                     -> True
+--        Cons {n1:Nat} {5~n1+1} (x :: a) (xs :: Vec n1 a) -> False
+--
+-- This function solves 'n1' and replaces every occurrence with its solution. A
+-- very limited number of solutions are currently recognized: only adds (such
+-- as in the example) will be solved.
+elimExistentials :: HasCallStack => NormRewrite
+elimExistentials (TransformContext is0 _) (Case scrut altsTy alts0) = do
+  tcm <- Lens.view tcCache
+  alts1 <- traverse (go is0 tcm) alts0
+  caseOneAlt (Case scrut altsTy alts1)
+ where
+    -- Eliminate free type variables if possible
+    go :: InScopeSet -> TyConMap -> Alt -> NormalizeSession Alt
+    go is2 tcm alt@(DataPat dc exts0 xs0, term0) =
+      case solveNonAbsurds tcm (mkVarSet exts0) (altEqs tcm alt) of
+        -- No equations solved:
+        [] -> return alt
+        -- One or more equations solved:
+        sols ->
+          changed =<< go is2 tcm (DataPat dc exts1 xs1, term1)
+          where
+            -- Substitute solution in existentials and applied types
+            is3   = extendInScopeSetList is2 exts0
+            xs1   = map (substTyInVar (extendTvSubstList (mkSubst is3) sols)) xs0
+            exts1 = substInExistentialsList is2 exts0 sols
+
+            -- Substitute solution in term.
+            is4       = extendInScopeSetList is3 xs1
+            subst     = extendTvSubstList (mkSubst is4) sols
+            term1     = substTm "Replacing tyVar due to solved eq" subst term0
```

Comment:
I'm not going to comment on style too much, but this is just weird :-)

Reply:
Yes, I missed those aligned `=`. Spend enough time reading these files and your eyes slowly glaze over